File: src/main/java/com/hileco/cortex/primitives/heavy/BackedVmMap.java
package com.hileco.cortex.primitives.heavy;


import com.hileco.vm.primitives.VmMap;

import java.util.HashMap;
import java.util.Map;

public class BackedVmMap<K, V> implements VmMap<K, V, BackedVmMap<K, V>> {
    private Map<K, V> backingMap;

    public BackedVmMap() {
        this.backingMap = new HashMap<>();
    }

    private BackedVmMap(Map<K, V> backingMap) {
        this.backingMap = new HashMap<>(backingMap);
    }

    public void close() {
    }

    public BackedVmMap<K, V> copy() {
        return new BackedVmMap<>(backingMap);
    }

    public V get(K key) {
        return backingMap.get(key);
    }

    public void put(K key, V value) {
        backingMap.put(key, value);
    }

    public void remove(K key) {
        backingMap.remove(key);
    }
}

File: cortex-data/src/main/kotlin/com/hileco/cortex/data/backed/BackedVmMap.kt
package com.hileco.cortex.data.backed


import com.hileco.cortex.data.base.BaseVmMap
import java.util.*

class BackedVmMap<K, V> : BaseVmMap<K, V, BackedVmMap<K, V>> {
    private var backingMap: MutableMap<K, V>

    constructor() {
        this.backingMap = HashMap()
    }

    private constructor(backingMap: MutableMap<K, V>) {
        this.backingMap = HashMap(backingMap)
    }

    override fun close() {}

    override fun copy(): BackedVmMap<K, V> {
        return BackedVmMap(backingMap)
    }

    override fun get(key: K): V? {
        return backingMap[key]
    }

    override fun set(key: K, value: V) {
        backingMap[key] = value
    }

    override fun remove(key: K) {
        backingMap.remove(key)
    }

    override fun keySet(): Set<K> {
        return backingMap.keys
    }
}
